packages/audiofileplayer/android/src/main/java/com/google/flutter/plugins/audiofileplayer/AudiofileplayerPlugin.java

```diff
@@ -112,37 +121,103 @@
   // Used when defining an Intent from a custom media button.
   public static final String CUSTOM_MEDIA_BUTTON_EXTRA_KEY = "customMediaButton";
 
-  private final Registrar registrar;
-  private final Map<String, ManagedMediaPlayer> mediaPlayers;
-  private final MethodChannel methodChannel;
+  private Registrar registrar;
+  private Activity activity;
+  private Map<String, ManagedMediaPlayer> mediaPlayers;
+  private MethodChannel methodChannel;
+  private Context context;
+  private ActivityPluginBinding activityPluginBinding;
 
   private MediaBrowserCompat mediaBrowser;
   private MediaControllerCompat mediaController;
 
   public static void registerWith(Registrar registrar) {
-    final MethodChannel methodChannel = new MethodChannel(registrar.messenger(), CHANNEL);
-    final AudiofileplayerPlugin instance = new AudiofileplayerPlugin(registrar, methodChannel);
-    methodChannel.setMethodCallHandler(instance);
+    AudiofileplayerPlugin instance = new AudiofileplayerPlugin();
+    instance.registrar = registrar;
+    instance.activity = registrar.activity();
+    instance.initInstance(registrar.messenger(), registrar.context());
+    instance.registerLifecycleCallbacks(instance.activity);
   }
 
-  private AudiofileplayerPlugin(Registrar registrar, MethodChannel methodChannel) {
-    this.registrar = registrar;
-    this.methodChannel = methodChannel;
-    this.mediaPlayers = new HashMap<>();
-    LifecycleCallbacks callbacks = new LifecycleCallbacks(this, registrar.activity().hashCode());
-    registrar.activity().getApplication().registerActivityLifecycleCallbacks(callbacks);
+  private void registerLifecycleCallbacks(Activity activity) {
+    LifecycleCallbacks callbacks = new LifecycleCallbacks(this, activity.hashCode());
+    activity.getApplication().registerActivityLifecycleCallbacks(callbacks);
+  }
 
-    // Set up MediaBrowser to connect to AudiofileplayerService. Service will be started on first
-    // playback of a background asset.
-    Context context = registrar.activeContext();
+  private void initInstance(BinaryMessenger messenger, Context context) {
+    this.context = context;
+    methodChannel = new MethodChannel(messenger, CHANNEL);
+    methodChannel.setMethodCallHandler(this);
+    mediaPlayers = new HashMap<>();
+    Context activeContext = activeContext();
     mediaBrowser =
         new MediaBrowserCompat(
-            context,
-            new ComponentName(context, AudiofileplayerService.class),
+            activeContext,
+            new ComponentName(activeContext, AudiofileplayerService.class),
             connectionCallback,
             null);
   }
 
+  private Activity activity() {
+    return activityPluginBinding != null ? activityPluginBinding.getActivity() : activity;
+  }
+
+  private Context activeContext() {
```

Comment:
Then looks like you don't need this. But I'm not sure the intention of using activeContext. You can wait for other reviewer's comment.

Reply:
See above.

Reply:
Once you migrated apps to v2, this method doesn't make sense anymore. Maybe v2 removed activeContext for a reason. It may not be necessary.

Reply:
Yes I agree. this one is only for the sake of v1. Do you suggest we use context directly here? For v2 that makes perfect sense but I am not sure if that introduces something on v1.